File: src/main/java/net/vulcanmc/vulcaneconomy/VulcanEconomy.java
package net.vulcanmc.vulcaneconomy;


import com.sk89q.squirrelid.Profile;
import com.sk89q.squirrelid.cache.SQLiteCache;
import com.sk89q.squirrelid.resolver.HttpRepositoryService;
import com.sk89q.squirrelid.resolver.ProfileService;
import kong.unirest.Unirest;
import net.vulcanmc.vulcaneconomy.commands.*;
import net.vulcanmc.vulcaneconomy.listeners.PlayerListener;
import net.vulcanmc.vulcaneconomy.rest.Accounts;
import net.vulcanmc.vulcaneconomy.rest.Currencies;
import net.vulcanmc.vulcaneconomy.rest.Currency;
import net.vulcanmc.vulcaneconomy.vault.Economy_VulcanEco;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;
import org.bukkit.plugin.PluginDescriptionFile;
import org.bukkit.plugin.ServicePriority;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.plugin.java.JavaPluginLoader;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.UUID;

//import static net.vulcanmc.vulcaneconomy.vault.ScoreboardLoader.load;

public class VulcanEconomy extends JavaPlugin{
    private static VulcanEconomy plugin;
    private static String apiURL;
    private static UUID serverid;
    private static ProfileService resolver;
    private static String password;
    private static String username;

    private SQLiteCache cache;
    private String apiUser;
    private String apiPass;
    //private RequestsQueue queue;
    private Currencies currencies = new Currencies();
    private Accounts accounts;// = new Accounts();


    private static UUID serverID;

//For use in MockBukkit
    public VulcanEconomy()
    {
        super();
    }

    protected VulcanEconomy(JavaPluginLoader loader, PluginDescriptionFile description, File dataFolder, File file)
    {
        super(loader, description, dataFolder, file);
    }

    public static VulcanEconomy getPlugin() {
        return plugin;
    }

    public static String getApiURL() {
        return apiURL;
    }

    public static UUID getServerid() {
        return serverid;
    }
    public static String getUsername() {
        return username;
    }
    public static String getPassword() {
        return password;
    }

    public static ProfileService getResolver() {
        return resolver;
    }

 /*   public RequestsQueue getQueue() {
        return queue;
    }
*/


    @Override

    public void onEnable() {
        plugin = this;

      //  this.queue = new RequestsQueue();
        this.saveDefaultConfig();
        getConfig().options().copyDefaults(true);
        //remember to disable commands in essentials config.
        this.getCommand("balance").setExecutor(new Balance(this));
        this.getCommand("balancetop").setExecutor(new BalanceTop(this));
        this.getCommand("pay").setExecutor(new Pay(this));
        this.getCommand("economy").setExecutor(new Economy(this));
        this.getCommand("transactions").setExecutor(new Transactions(this));

        this.getServer().getPluginManager().registerEvents(new PlayerListener(this), this);
        String serverIDstr = this.getConfig().getString("server-id");
        this.serverID = UUID.fromString(serverIDstr);
        this.apiURL = this.getConfig().getString("api-url");
        this.apiUser = this.getConfig().getString("api-username");
        this.apiPass = this.getConfig().getString("api-password");
        username = apiUser;
        password = apiPass;
        setupVault();
        this.resolver = HttpRepositoryService.forMinecraft();
        File file = new File("uuidcache.sqlite");

        if(getPlugin().getServer().getPluginManager().getPlugin("ScoreboardStats") != null) {
            //load(this);
        }

accounts = new Accounts();


    }
    public UUID getUUIDIfExists(String playername) {
        if(Bukkit.getPlayer(playername) != null)
        {
            return Bukkit.getPlayer(playername).getUniqueId();
        }
        /*

        Integer playerid;
        UUID uuid = null;
        //this.player = (Player) player;
        try {
            JsonNode response = Unirest.get(VulcanEconomy.getApiURL() + "players/").basicAuth(this.getApiUser(), this.getApiPass()).asJson().getBody();

            JSONArray data = response.getObject().getJSONArray("data");

            for (int i = 0; i < data.length(); i++) {
                JSONObject object = data.getJSONObject(i);
                String usernameobject = object.getString("username");
                if(usernameobject.equals(playername)) {
                    playerid = object.getInt("id");
                    uuid = UUID.fromString(object.getString("uuid"));
                    return uuid;
                }
            }
        } catch (UnirestException e) {
            e.printStackTrace();
        }*/

        return null;
    }
    public UUID getUUID(String playername) {
        if(Bukkit.getPlayer(playername) != null)
        {
            return Bukkit.getPlayer(playername).getUniqueId();
        }

        Integer playerid;
        UUID uuid = getUUIDIfExists(playername);
            if(uuid == null) {
                try {
                    Profile profile = this.getResolver().findByName(playername);
                    if(profile != null) {
                        uuid = profile.getUniqueId();
                        return uuid;
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
            return Bukkit.getOfflinePlayer(playername).getUniqueId();
            //lookup with usernmae instead of uuid. fall back with mojang api
    }
    @Override
    public void onDisable() {
        for(Player player : getServer().getOnlinePlayers()){
            accounts.removeAccountsFromCache(player.getUniqueId());
        }
        Unirest.shutDown();

        plugin = null;
    }

    private void setupVault() {
        Plugin vault = getServer().getPluginManager().getPlugin("Vault");

        if (vault == null) {
            return;
        }

        getServer().getServicesManager().register(net.milkbowl.vault.economy.Economy.class, new Economy_VulcanEco(this), this, ServicePriority.Highest);
        //getServer().getServicesManager().unregister(net.milkbowl.vault.economy.Economy.class, net.milkbowl.vault.economy.plugins.Economy_Essentials.class);
    }


    public SQLiteCache getCache() {
        return cache;
    }

    public String getApiUser() {
        return apiUser;
    }

    public String getApiPass() {
        return apiPass;
    }

    public static UUID getServerID() {
        return serverID;
    }

    public Accounts getAccounts() {
        return accounts;
    }

    public Currencies getCurrencies() {
        return currencies;
    }
}

File: src/main/java/net/vulcanmc/vulcaneconomy/VulcanEconomy.kt
package net.vulcanmc.vulcaneconomy


import net.vulcanmc.vulcaneconomy.commands.*
import net.vulcanmc.vulcaneconomy.listeners.PlayerListener
import net.vulcanmc.vulcaneconomy.rest.Accounts
import net.vulcanmc.vulcaneconomy.rest.Currencies
import net.vulcanmc.vulcaneconomy.vault.Economy_VulcanEco
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.plugin.PluginDescriptionFile
import org.bukkit.plugin.ServicePriority
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.plugin.java.JavaPluginLoader
import org.enginehub.squirrelid.resolver.HttpRepositoryService
import org.enginehub.squirrelid.resolver.ProfileService

import java.io.File
import java.io.IOException
import java.util.UUID

//import static net.vulcanmc.vulcaneconomy.vault.ScoreboardLoader.load;

class VulcanEconomy : JavaPlugin {
  //  private var apiURL: String = ""
    private var serverID: UUID? = null
    private lateinit var resolver : ProfileService
    var apiUser: String = ""
        private set
    var apiPass: String = ""
        private set
    lateinit var uuid : UUID;
    //private RequestsQueue queue;
    val currencies = Currencies()
    var accounts: Accounts? = null
        private set// = new Accounts();
    val prefix: String
        get() = ChatColor.translateAlternateColorCodes('&', config.getString("prefix"))

    //For use in MockBukkit
    constructor() : super() {}

    protected constructor(loader: JavaPluginLoader, description: PluginDescriptionFile, dataFolder: File, file: File) : super(loader, description, dataFolder, file) {}


    fun reload() {
        val serverIDstr = this.config.getString("server-id")
        this.serverID = UUID.fromString(serverIDstr)
        apiURL = this.config.getString("api-url")
        this.apiUser = this.config.getString("api-username")
        this.apiPass = this.config.getString("api-password")
        username = this.apiUser
        password = this.apiPass
        accounts!!.cache
        accounts!!.reloadCache()

    }

    override fun onEnable() {
        plugin = this

        //  this.queue = new RequestsQueue();
        this.saveDefaultConfig()
        config.options().copyDefaults(true)
        //remember to disable commands in essentials config.
        this.getCommand("balance").executor = Balance(this)
        this.getCommand("balancetop").executor = BalanceTop(this)
        this.getCommand("pay").executor = Pay(this)
        this.getCommand("economy").executor = Economy(this)
        this.getCommand("transactions").executor = Transactions(this)

        this.server.pluginManager.registerEvents(PlayerListener(this), this)
        setupVault()
        this.resolver = HttpRepositoryService.forMinecraft()

        val file = File("uuidcache.sqlite")

        if (plugin!!.server.pluginManager.getPlugin("ScoreboardStats") != null) {
            //load(this);
        }

        accounts = Accounts()
        reload()


    }

    fun getUUIDIfExists(playerName: String): UUID? {
        return if (Bukkit.getPlayer(playerName) != null) {
            Bukkit.getPlayer(playerName).uniqueId
        } else null
    }

    fun getUUID(playerName: String): UUID? {
        if (Bukkit.getPlayer(playerName) != null) {
            return Bukkit.getPlayer(playerName).uniqueId
        }

        var uuid = getUUIDIfExists(playerName)
        if (uuid == null) {
            try {
                val profile = this.resolver!!.findByName(playerName)
                if (profile != null) {
                    uuid = profile!!.getUniqueId()
                    return uuid
                }
            } catch (e: IOException) {
                e.printStackTrace()
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }

        } else {
            return uuid;
        }
        return Bukkit.getOfflinePlayer(playerName).uniqueId //todo: check if this is ever reached?
    }

    override fun onDisable() {
        for (player in server.onlinePlayers) {
            accounts!!.removeAccountsFromCache(player.uniqueId)
        }

        plugin = null
    }

    private fun setupVault() {
        val vault = server.pluginManager.getPlugin("Vault") ?: return

        server.servicesManager.register(net.milkbowl.vault.economy.Economy::class.java, Economy_VulcanEco(this), this, ServicePriority.Highest)
        //getServer().getServicesManager().unregister(net.milkbowl.vault.economy.Economy.class, net.milkbowl.vault.economy.plugins.Economy_Essentials.class);
    }

    companion object {
        var plugin: VulcanEconomy? = null
            private set

        var password: String = ""
            private set
        var username: String = ""
            private set

        var apiURL : String = ""
            private set
        var serverID: UUID? = null
            private set
    }
}
